javascripts/discourse/components/discourse-category-banners.js

```diff
@@ -0,0 +1,110 @@
+import Category from "discourse/models/category";
+import Component from "@glimmer/component";
+import { tracked } from "@glimmer/tracking";
+import { action } from "@ember/object";
+import { inject as service } from "@ember/service";
+import { htmlSafe } from "@ember/template";
+
+export default class DiscourseCategoryBanners extends Component {
+  @service router;
+  @service site;
+  @tracked category = null;
+
+  constructor() {
+    super(...arguments);
+    this.router.on("routeDidChange", this, this.getCategory);
+  }
+
+  willDestroy() {
+    super.willDestroy(...arguments);
+    this.router.off("routeDidChange", this, this.getCategory);
+  }
+
+  get shouldShow() {
+    const router = this.router;
+    const route = router.currentRoute;
+    if (
+      route &&
+      route.params &&
+      route.params.hasOwnProperty("category_slug_path_with_id")
+    ) {
+      const categories = {};
+
+      settings.categories.split("|").forEach((item) => {
+        item = item.split(":");
+
+        if (item[0]) {
+          categories[item[0]] = item[1] || "all";
+        }
+      });
+
+      const isException = settings.exceptions
+        .split("|")
+        .filter(Boolean)
+        .map((value) => value.toLowerCase())
+        .includes(this.category.name.toLowerCase());
+
+      const isTarget =
+        Object.keys(categories).length === 0 ||
+        categories[this.category.name] === "all" ||
+        categories[this.category.name] === "no_sub" ||
+        (this.category.parentCategory &&
+          (categories[this.category.parentCategory.name] === "all" ||
+            categories[this.category.parentCategory.name] === "only_sub"));
+
+      const hideMobile = !settings.show_mobile && this.site.mobileView;
+      const isSubCategory =
+        !settings.show_subcategory && this.category.parentCategory;
+      const hasNoCategoryDescription =
+        settings.hide_if_no_description && !this.category.description_text;
+
+      if (
+        isTarget &&
+        !isException &&
+        !hasNoCategoryDescription &&
+        !isSubCategory &&
+        !hideMobile
+      ) {
+        document.body.classList.add("category-header");
+      } else {
+        document.body.classList.remove("category-header");
+      }
+    }
+  }
+
+  get safeStyle() {
+    return htmlSafe(
+      `background-color: #${this.category.color}; color: #${this.category.text_color};`
+    );
+  }
+
+  get rerenderWidget() {
+    // forces the category icon widget to rerender when the category changes
+    const router = this.router;
+    const route = router.currentRoute;
+
+    return (
+      route &&
+      route.params &&
+      route.params.hasOwnProperty("category_slug_path_with_id")
+    );
+  }
```

Comment:
Does this actually help the widget to re-render correctly? In theory, if this continues to return true, a handlebars `{{#if` won't cause anything inside to rerender 🤔

Reply:
Ah I see, that makes sense. But then if it's relying on the intermediate loading route, it's probably not going to work alongside the loading-slider theme component?

I think the best solution is to define a dedicated component in the category-icons theme component like


```js
export default class CategoryIcon extends MountWidget {
  widget = "category-icon";
  
  buildArgs(){
    return { category: this.category }
  }
  
  didReceiveAttrs(){
    super.didReceiveAttrs();
    this.queueRerender();
  }
}
```

and then we can use it here like `<CategoryIcon @category=this.category />`. It also means that if/when the category-icons theme component is glimmer-ified, it'll be transparent to themes that lean on it.